server-common/src/main/java/org/apache/kafka/server/common/serialization/AbstractApiMessageSerde.java

```diff
@@ -69,16 +76,42 @@ public void write(ApiMessageAndVersion data,
     @Override
     public ApiMessageAndVersion read(Readable input,
                                      int size) {
-        short frameVersion = (short) input.readUnsignedVarint();
+        short frameVersion;
+        try {
+            frameVersion = unsignedIntToShort(input.readUnsignedVarint(), "frame version");
+        } catch (Exception e) {
+            throw new MetadataParseException("Failed to read variable-length " +
+                    "frame version number: " + e.getClass().getSimpleName() + ": " + e.getMessage());
+        }
         if (frameVersion != DEFAULT_FRAME_VERSION) {
             throw new SerializationException("Could not deserialize metadata record due to unknown frame version "
```

Comment:
Should we use the parse exception consistently?

Reply:
Done!